Should @types/request be in dependencies, rather than devDependencies?

When I compile my typescript project (which uses gcp-metadata) and I do not have the package @types/request, I get the following error.

```
1 /// <reference types="request" />
                        ~~~~~~~

node_modules/gcp-metadata/index.d.ts(1,23): error TS2688: Cannot find type definition file for 'request'.


2 import * as _request from 'request';
                            ~~~~~~~~~

node_modules/gcp-metadata/index.d.ts(2,27): error TS7016: Could not find a declaration file for module 'request'. '/usr/local/google/home/nolanmar/NODE/profile-collection/cloud-profiler-nodejs/node_modules/request/index.js' implicitly has an 'any' type.
  Try `npm install @types/request` if it exists or add a new declaration (.d.ts) file containing `declare module 'request';`
```

If I add @types/request to the devDependenvies for my project, I do not get this error.

--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,4 @@
 node_modules
 build
 .vscode
+package-lock.json

--- a/README.md
+++ b/README.md
@@ -26,7 +26,7 @@ gcpMetadata.instance('hostname', function(err, response, metadata) {
 ```js
 gcpMetadata.instance({
   property: 'tags',
-  qs: { alt: 'text' }
+  params: { alt: 'text' }
 }, function(err, response, metadata) {
   // Tags as newline-delimited list
 })

--- a/package.json
+++ b/package.json
@@ -34,7 +34,8 @@
   "license": "MIT",
   "dependencies": {
     "axios": "^0.17.1",
-    "extend": "^3.0.1"
+    "extend": "^3.0.1",
+    "retry-axios": "0.0.1"
   },
   "devDependencies": {
     "@types/extend": "^3.0.0",

--- a/package.json
+++ b/package.json
@@ -33,19 +33,15 @@
   "author": "Stephen Sawchuk",
   "license": "MIT",
   "dependencies": {
-    "extend": "^3.0.0",
-    "retry-request": "^3.1.0"
+    "axios": "^0.17.1"
   },
   "devDependencies": {
-    "@types/extend": "^3.0.0",
     "@types/mocha": "^2.2.43",
+    "@types/nock": "^9.1.2",
     "@types/node": "^8.0.31",
-    "@types/proxyquire": "^1.3.28",
-    "@types/request": "^2.0.3",
-    "clang-format": "^1.1.0",
-    "gts": "^0.3.1",
+    "gts": "^0.5.1",
     "mocha": "^3.2.0",
-    "proxyquire": "^1.7.10",
+    "nock": "^9.1.6",
     "typescript": "^2.5.3"
   },
   "engines": {

--- a/package.json
+++ b/package.json
@@ -35,7 +35,7 @@
   "dependencies": {
     "axios": "^0.17.1",
     "extend": "^3.0.1",
-    "retry-axios": "0.2.0"
+    "retry-axios": "0.3.0"
   },
   "devDependencies": {
     "@types/extend": "^3.0.0",

--- a/package.json
+++ b/package.json
@@ -33,9 +33,11 @@
   "author": "Stephen Sawchuk",
   "license": "MIT",
   "dependencies": {
-    "axios": "^0.17.1"
+    "axios": "^0.17.1",
+    "extend": "^3.0.1"
   },
   "devDependencies": {
+    "@types/extend": "^3.0.0",
     "@types/mocha": "^2.2.43",
     "@types/nock": "^9.1.2",
     "@types/node": "^8.0.31",

--- a/package.json
+++ b/package.json
@@ -35,7 +35,7 @@
   "dependencies": {
     "axios": "^0.17.1",
     "extend": "^3.0.1",
-    "retry-axios": "0.0.1"
+    "retry-axios": "0.2.0"
   },
   "devDependencies": {
     "@types/extend": "^3.0.0",

--- a/src/index.ts
+++ b/src/index.ts
@@ -12,10 +12,10 @@ export type Callback =
 
 // Accepts an options object passed from the user to the API.  In the
 // previous version of the API, it referred to a `Request` options object.
-// Now it refers to an Axiox Request Config object.  This is here to help
+// Now it refers to an Axios Request Config object.  This is here to help
 // ensure users don't pass invalid options when they upgrade from 0.4 to 0.5.
 // tslint:disable-next-line no-any
-function validate(options: any) {
+function validate(options: any, callback: (err?: Error) => void) {
   const vpairs = [
     {invalid: 'uri', expected: 'url'}, {invalid: 'json', expected: 'data'},
     {invalid: 'qs', expected: 'params'}
@@ -25,9 +25,10 @@ function validate(options: any) {
       const e = `'${
           pair.invalid}' is not a valid configuration option. Please use '${
           pair.expected}' instead. This library is using Axios for requests. Please see https://github.com/axios/axios to learn more about the valid request options.`;
-      throw new Error(e);
+      return callback(new Error(e));
     }
   }
+  callback();
 }
 
 export function _buildMetadataAccessor(type: string) {
@@ -51,44 +52,34 @@ export function _buildMetadataAccessor(type: string) {
       property = '/' + options.property;
     }
 
-    try {
-      validate(options);
-    } catch (e) {
-      if (callback) {
-        callback!(e);
-      } else {
-        throw e;
+    validate(options, err => {
+      if (err) {
+        return callback!(err);
       }
-    }
 
-    const ax = axios.create();
-    rax.attach(ax);
-    const baseOpts = {
-      url: `${BASE_URL}/${type}${property}`,
-      headers: {'Metadata-Flavor': 'Google'}
-    };
-    const reqOpts = extend(true, baseOpts, options);
-    delete reqOpts.property;
+      const ax = axios.create();
+      rax.attach(ax);
+      const baseOpts = {
+        url: `${BASE_URL}/${type}${property}`,
+        headers: {'Metadata-Flavor': 'Google'}
+      };
+      const reqOpts = extend(true, baseOpts, options);
+      delete (reqOpts as {property: string}).property;
 
-    return ax.request(reqOpts)
-        .then(res => {
-          if (callback) {
-            callback(null, res, res.data);
-          }
-        })
-        .catch((err: AxiosError) => {
-          let e: Error = err;
-          if (err.response && !err.response.data) {
-            e = new Error('Invalid response from metadata service');
-          } else if (err.code !== '200') {
-            e = new Error('Unsuccessful response status code');
-          }
-          if (callback) {
-            callback(e);
-          } else {
-            throw e;
-          }
-        });
+      ax.request(reqOpts)
+          .then(res => {
+            callback!(null, res, res.data);
+          })
+          .catch((err: AxiosError) => {
+            let e: Error = err;
+            if (err.response && !err.response.data) {
+              e = new Error('Invalid response from metadata service');
+            } else if (err.code !== '200') {
+              e = new Error('Unsuccessful response status code');
+            }
+            callback!(e);
+          });
+    });
   };
 }
 

--- a/src/index.ts
+++ b/src/index.ts
@@ -62,7 +62,8 @@ export function _buildMetadataAccessor(type: string) {
     rax.attach(ax);
     const baseOpts = {
       url: `${BASE_URL}/${type}${property}`,
-      headers: {'Metadata-Flavor': 'Google'}
+      headers: {'Metadata-Flavor': 'Google'},
+      raxConfig: {noResponseRetries: 0}
     };
     const reqOpts = extend(true, baseOpts, options);
     delete (reqOpts as {property: string}).property;

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,13 +1,11 @@
-import * as extend from 'extend';
-import * as _request from 'request';  // for types only
-import * as request from 'retry-request';
+import axios, {AxiosError, AxiosRequestConfig, AxiosResponse} from 'axios';
 
 const BASE_URL = 'http://metadata.google.internal/computeMetadata/v1';
 
-export type Options = _request.CoreOptions&{property?: string, uri?: string};
+export type Options = AxiosRequestConfig&{property?: string, uri?: string};
 
 export type Callback =
-    (error: NodeJS.ErrnoException|null, response?: _request.RequestResponse,
+    (error: NodeJS.ErrnoException|null, response?: AxiosResponse<string>,
      metadataProp?: string) => void;
 
 export function _buildMetadataAccessor(type: string) {
@@ -27,30 +25,31 @@ export function _buildMetadataAccessor(type: string) {
       property = '/' + options.property;
     }
 
-    const reqOpts = extend(
-                        true, {
-                          uri: BASE_URL + '/' + type + property,
+    const reqOpts = Object.assign(
+                        {
+                          url: `${BASE_URL}/${type}${property}`,
                           headers: {'Metadata-Flavor': 'Google'}
                         },
-                        options) as _request.Options &
+                        options) as AxiosRequestConfig &
         {property?: string};
     delete reqOpts.property;
 
-    const retryRequestOpts = {noResponseRetries: 0};
-
-    return request(reqOpts, retryRequestOpts, (err, res, body) => {
-      if (callback) {  // for type safety; this should already always be true
-        if (err) {
-          callback(err);
-        } else if (!res) {
-          callback(new Error('Invalid response from metadata service'));
-        } else if (res.statusCode !== 200) {
-          callback(new Error('Unsuccessful response status code'), res);
-        } else {
-          callback(null, res, body);
-        }
-      }
-    });
+    axios(reqOpts)
+        .then(res => {
+          if (callback) callback(null, res, res.data);
+        })
+        .catch((err: AxiosError) => {
+          if (callback) {
+            if (err.response && !err.response.data) {
+              callback(new Error('Invalid response from metadata service'));
+            } else if (err.code !== '200') {
+              callback(
+                  new Error('Unsuccessful response status code'), err.response);
+            } else {
+              callback(err);
+            }
+          }
+        });
   };
 }
 

--- a/src/index.ts
+++ b/src/index.ts
@@ -4,7 +4,8 @@ import * as rax from 'retry-axios';
 
 const BASE_URL = 'http://metadata.google.internal/computeMetadata/v1';
 
-export type Options = AxiosRequestConfig&{property?: string, uri?: string};
+export type Options = AxiosRequestConfig&
+    {[index: string]: {} | string | undefined, property?: string, uri?: string};
 
 export type Callback =
     (error: NodeJS.ErrnoException|null, response?: AxiosResponse<string>,
@@ -14,8 +15,7 @@ export type Callback =
 // previous version of the API, it referred to a `Request` options object.
 // Now it refers to an Axios Request Config object.  This is here to help
 // ensure users don't pass invalid options when they upgrade from 0.4 to 0.5.
-// tslint:disable-next-line no-any
-function validate(options: any) {
+function validate(options: Options) {
   const vpairs = [
     {invalid: 'uri', expected: 'url'}, {invalid: 'json', expected: 'data'},
     {invalid: 'qs', expected: 'params'}
@@ -82,6 +82,8 @@ export function _buildMetadataAccessor(type: string) {
         });
   };
 }
+
+
 
 export const instance = _buildMetadataAccessor('instance');
 export const project = _buildMetadataAccessor('project');

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,5 +1,6 @@
 import axios, {AxiosError, AxiosRequestConfig, AxiosResponse} from 'axios';
 import * as extend from 'extend';
+import * as rax from 'retry-axios';
 
 const BASE_URL = 'http://metadata.google.internal/computeMetadata/v1';
 
@@ -29,10 +30,13 @@ function validate(options: any) {
   }
 }
 
-
 export function _buildMetadataAccessor(type: string) {
   return function metadataAccessor(
-      options: string|Options|Callback, callback?: Callback) {
+      options?: string|Options|Callback, callback?: Callback) {
+    if (!options) {
+      options = {};
+    }
+
     if (typeof options === 'function') {
       callback = options;
       options = {};
@@ -50,21 +54,27 @@ export function _buildMetadataAccessor(type: string) {
     try {
       validate(options);
     } catch (e) {
-      return callback!(e);
+      if (callback) {
+        callback!(e);
+      } else {
+        throw e;
+      }
     }
 
-    const reqOpts = extend(
-                        true, {
-                          url: `${BASE_URL}/${type}${property}`,
-                          headers: {'Metadata-Flavor': 'Google'}
-                        },
-                        options) as AxiosRequestConfig &
-        {property?: string};
+    const ax = axios.create();
+    rax.attach(ax);
+    const baseOpts = {
+      url: `${BASE_URL}/${type}${property}`,
+      headers: {'Metadata-Flavor': 'Google'}
+    };
+    const reqOpts = extend(true, baseOpts, options);
     delete reqOpts.property;
 
-    axios(reqOpts)
+    return ax.request(reqOpts)
         .then(res => {
-          callback!(null, res, res.data);
+          if (callback) {
+            callback(null, res, res.data);
+          }
         })
         .catch((err: AxiosError) => {
           let e: Error = err;
@@ -73,7 +83,11 @@ export function _buildMetadataAccessor(type: string) {
           } else if (err.code !== '200') {
             e = new Error('Unsuccessful response status code');
           }
-          return callback!(e);
+          if (callback) {
+            callback(e);
+          } else {
+            throw e;
+          }
         });
   };
 }

--- a/src/index.ts
+++ b/src/index.ts
@@ -15,7 +15,7 @@ export type Callback =
 // Now it refers to an Axios Request Config object.  This is here to help
 // ensure users don't pass invalid options when they upgrade from 0.4 to 0.5.
 // tslint:disable-next-line no-any
-function validate(options: any, callback: (err?: Error) => void) {
+function validate(options: any) {
   const vpairs = [
     {invalid: 'uri', expected: 'url'}, {invalid: 'json', expected: 'data'},
     {invalid: 'qs', expected: 'params'}
@@ -25,15 +25,15 @@ function validate(options: any, callback: (err?: Error) => void) {
       const e = `'${
           pair.invalid}' is not a valid configuration option. Please use '${
           pair.expected}' instead. This library is using Axios for requests. Please see https://github.com/axios/axios to learn more about the valid request options.`;
-      return callback(new Error(e));
+      return new Error(e);
     }
   }
-  callback();
+  return null;
 }
 
 export function _buildMetadataAccessor(type: string) {
   return function metadataAccessor(
-      options?: string|Options|Callback, callback?: Callback) {
+      options: string|Options|Callback, callback?: Callback) {
     if (!options) {
       options = {};
     }
@@ -52,34 +52,34 @@ export function _buildMetadataAccessor(type: string) {
       property = '/' + options.property;
     }
 
-    validate(options, err => {
-      if (err) {
-        return setImmediate(callback!, err);
-      }
+    const err = validate(options);
+    if (err) {
+      setImmediate(callback!, err);
+      return;
+    }
 
-      const ax = axios.create();
-      rax.attach(ax);
-      const baseOpts = {
-        url: `${BASE_URL}/${type}${property}`,
-        headers: {'Metadata-Flavor': 'Google'}
-      };
-      const reqOpts = extend(true, baseOpts, options);
-      delete (reqOpts as {property: string}).property;
+    const ax = axios.create();
+    rax.attach(ax);
+    const baseOpts = {
+      url: `${BASE_URL}/${type}${property}`,
+      headers: {'Metadata-Flavor': 'Google'}
+    };
+    const reqOpts = extend(true, baseOpts, options);
+    delete (reqOpts as {property: string}).property;
 
-      ax.request(reqOpts)
-          .then(res => {
-            callback!(null, res, res.data);
-          })
-          .catch((err: AxiosError) => {
-            let e: Error = err;
-            if (err.response && !err.response.data) {
-              e = new Error('Invalid response from metadata service');
-            } else if (err.code !== '200') {
-              e = new Error('Unsuccessful response status code');
-            }
-            callback!(e);
-          });
-    });
+    ax.request(reqOpts)
+        .then(res => {
+          callback!(null, res, res.data);
+        })
+        .catch((err: AxiosError) => {
+          let e: Error = err;
+          if (err.response && !err.response.data) {
+            e = new Error('Invalid response from metadata service');
+          } else if (err.code !== '200') {
+            e = new Error('Unsuccessful response status code');
+          }
+          callback!(e);
+        });
   };
 }
 

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,4 +1,5 @@
 import axios, {AxiosError, AxiosRequestConfig, AxiosResponse} from 'axios';
+import * as extend from 'extend';
 
 const BASE_URL = 'http://metadata.google.internal/computeMetadata/v1';
 
@@ -7,6 +8,27 @@ export type Options = AxiosRequestConfig&{property?: string, uri?: string};
 export type Callback =
     (error: NodeJS.ErrnoException|null, response?: AxiosResponse<string>,
      metadataProp?: string) => void;
+
+// Accepts an options object passed from the user to the API.  In the
+// previous version of the API, it referred to a `Request` options object.
+// Now it refers to an Axiox Request Config object.  This is here to help
+// ensure users don't pass invalid options when they upgrade from 0.4 to 0.5.
+// tslint:disable-next-line no-any
+function validate(options: any) {
+  const vpairs = [
+    {invalid: 'uri', expected: 'url'}, {invalid: 'json', expected: 'data'},
+    {invalid: 'qs', expected: 'params'}
+  ];
+  for (const pair of vpairs) {
+    if (options[pair.invalid]) {
+      const e = `'${
+          pair.invalid}' is not a valid configuration option. Please use '${
+          pair.expected}' instead. This library is using Axios for requests. Please see https://github.com/axios/axios to learn more about the valid request options.`;
+      throw new Error(e);
+    }
+  }
+}
+
 
 export function _buildMetadataAccessor(type: string) {
   return function metadataAccessor(
@@ -25,8 +47,14 @@ export function _buildMetadataAccessor(type: string) {
       property = '/' + options.property;
     }
 
-    const reqOpts = Object.assign(
-                        {
+    try {
+      validate(options);
+    } catch (e) {
+      return callback!(e);
+    }
+
+    const reqOpts = extend(
+                        true, {
                           url: `${BASE_URL}/${type}${property}`,
                           headers: {'Metadata-Flavor': 'Google'}
                         },
@@ -36,19 +64,16 @@ export function _buildMetadataAccessor(type: string) {
 
     axios(reqOpts)
         .then(res => {
-          if (callback) callback(null, res, res.data);
+          callback!(null, res, res.data);
         })
         .catch((err: AxiosError) => {
-          if (callback) {
-            if (err.response && !err.response.data) {
-              callback(new Error('Invalid response from metadata service'));
-            } else if (err.code !== '200') {
-              callback(
-                  new Error('Unsuccessful response status code'), err.response);
-            } else {
-              callback(err);
-            }
+          let e: Error = err;
+          if (err.response && !err.response.data) {
+            e = new Error('Invalid response from metadata service');
+          } else if (err.code !== '200') {
+            e = new Error('Unsuccessful response status code');
           }
+          return callback!(e);
         });
   };
 }

--- a/src/index.ts
+++ b/src/index.ts
@@ -54,7 +54,7 @@ export function _buildMetadataAccessor(type: string) {
 
     validate(options, err => {
       if (err) {
-        return callback!(err);
+        return setImmediate(callback!, err);
       }
 
       const ax = axios.create();

--- a/test/index.test.ts
+++ b/test/index.test.ts
@@ -1,5 +1,7 @@
 import * as assert from 'assert';
+import * as extend from 'extend';
 import * as nock from 'nock';
+
 import * as gcpMetadata from '../src';
 
 const HOST = 'http://metadata.google.internal';
@@ -47,7 +49,7 @@ describe('gcpMetadata', () => {
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
     nock(HOST).get(`${PATH}/${TYPE}/${PROPERTY}`).query(QUERY).reply(200, {});
     getMetadata({property: PROPERTY, params: QUERY}, (err, res) => {
-      assert.equal(res!.config.params, QUERY);
+      assert.equal(JSON.stringify(res!.config.params), JSON.stringify(QUERY));
       assert.equal(res!.config.headers['Metadata-Flavor'], 'Google');
       assert.equal(res!.config.url, `${BASE_URL}/${TYPE}/${PROPERTY}`);
       done();
@@ -57,7 +59,7 @@ describe('gcpMetadata', () => {
   it('should extend the request options', (done) => {
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
     const options = {property: PROPERTY, headers: {'Custom-Header': 'Custom'}};
-    const originalOptions = Object.assign({}, options);
+    const originalOptions = extend(true, {}, options);
     nock(HOST).get(`${PATH}/${TYPE}/${PROPERTY}`).reply(200, {});
     getMetadata(options, (err, res) => {
       assert.equal(res!.config.url, `${BASE_URL}/${TYPE}/${PROPERTY}`);
@@ -101,6 +103,15 @@ describe('gcpMetadata', () => {
     nock(HOST).get(PATH).reply(418, {}, {'Metadata-Flavor': 'Google'});
     getMetadata((err) => {
       assert(err instanceof Error);
+      done();
+    });
+  });
+
+  it('should throw if request options are passed', (done) => {
+    // tslint:disable-next-line no-any
+    (gcpMetadata as any).instance({qs: {one: 'two'}}, (err: Error) => {
+      assert.notEqual(err, null);
+      assert(err.message.startsWith('\'qs\' is not a valid'));
       done();
     });
   });

--- a/test/index.test.ts
+++ b/test/index.test.ts
@@ -127,4 +127,14 @@ describe('gcpMetadata', () => {
       done();
     });
   });
+
+  it('should not retry on DNS errors', (done) => {
+    const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
+    nock(HOST).get(PATH).replyWithError({code: 'ETIMEDOUT'});
+    nock(HOST).get(PATH).reply(200, {});
+    getMetadata(err => {
+      assert(err instanceof Error);
+      done();
+    });
+  });
 });

--- a/test/index.test.ts
+++ b/test/index.test.ts
@@ -107,7 +107,19 @@ describe('gcpMetadata', () => {
     });
   });
 
+  it('should retry if the initial request fails', (done) => {
+    const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
+    nock(HOST).get(`${PATH}/${TYPE}`).reply(500);
+    nock(HOST).get(`${PATH}/${TYPE}`).reply(200, {});
+    getMetadata((err, res) => {
+      assert.equal(res!.config.url, `${BASE_URL}/${TYPE}`);
+      assert.equal(res!.config.headers['Metadata-Flavor'], 'Google');
+      done();
+    });
+  });
+
   it('should throw if request options are passed', (done) => {
+    const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
     // tslint:disable-next-line no-any
     (gcpMetadata as any).instance({qs: {one: 'two'}}, (err: Error) => {
       assert.notEqual(err, null);

--- a/test/index.test.ts
+++ b/test/index.test.ts
@@ -1,39 +1,20 @@
 import * as assert from 'assert';
-import * as extend from 'extend';
-import * as proxyquire from 'proxyquire';
-import * as _gcpMetadata from '../src';
+import * as nock from 'nock';
+import * as gcpMetadata from '../src';
 
-type GcpMetadata = typeof _gcpMetadata;
-type RetryRequest =
-    (reqOpts: Object, opts: Object,
-     callback: (err?: any, request?: any) => void) => void;
+const HOST = 'http://metadata.google.internal';
+const PATH = '/computeMetadata/v1';
+const BASE_URL = `${HOST}${PATH}`;
+const TYPE = 'type';
+const PROPERTY = 'property';
 
-// tslint:disable-next-line:no-empty
-const noop = () => {};
-
-const VALID_RESPONSE = {
-  statusCode: 200
-};
+function createNock() {
+  nock(HOST).get(PATH).reply(200, {});
+}
 
 describe('gcpMetadata', () => {
-  let cachedGcpMetadata: GcpMetadata;
-  let gcpMetadata: GcpMetadata;
-
-  let retryRequestOverride: RetryRequest|null;
-  let fakeRetryRequest = (...args: any[]) => {
-    return ((retryRequestOverride || noop) as Function)(...args);
-  };
-
-  before(() => {
-    cachedGcpMetadata =
-        proxyquire('../src/index.js', {'retry-request': fakeRetryRequest});
-  });
-
-  beforeEach(() => {
-    retryRequestOverride = null;
-    gcpMetadata =
-        proxyquire('../src/index.js', {'retry-request': fakeRetryRequest});
-    extend(gcpMetadata, cachedGcpMetadata);
+  afterEach(() => {
+    nock.cleanAll();
   });
 
   it('should create the correct accessors', () => {
@@ -42,79 +23,46 @@ describe('gcpMetadata', () => {
   });
 
   it('should access all the metadata properly', (done) => {
-    const BASE_URL = 'http://metadata.google.internal/computeMetadata/v1';
-    const TYPE = 'type';
-
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
-
-    retryRequestOverride = (reqOpts, opts, callback) => {
-      assert.deepEqual(
-          reqOpts,
-          {uri: BASE_URL + '/' + TYPE, headers: {'Metadata-Flavor': 'Google'}});
-      callback(null, VALID_RESPONSE);
-    };
-
-    getMetadata(done);
+    nock(HOST).get(`${PATH}/${TYPE}`).reply(200, {});
+    getMetadata((err, res) => {
+      assert.equal(res!.config.url, `${BASE_URL}/${TYPE}`);
+      assert.equal(res!.config.headers['Metadata-Flavor'], 'Google');
+      done();
+    });
   });
 
   it('should access a specific metadata property', (done) => {
-    const BASE_URL = 'http://metadata.google.internal/computeMetadata/v1';
-    const TYPE = 'type';
-    const PROPERTY = 'property';
-
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
-
-    retryRequestOverride = (reqOpts, opts, callback) => {
-      assert.deepEqual(reqOpts, {
-        uri: BASE_URL + '/' + TYPE + '/' + PROPERTY,
-        headers: {'Metadata-Flavor': 'Google'}
-      });
-      callback(null, VALID_RESPONSE);
-    };
-
-    getMetadata(PROPERTY, done);
+    nock(HOST).get(`${PATH}/${TYPE}/${PROPERTY}`).reply(200, {});
+    getMetadata(PROPERTY, (err, res) => {
+      assert.equal(res!.config.url, `${BASE_URL}/${TYPE}/${PROPERTY}`);
+      assert.equal(res!.config.headers['Metadata-Flavor'], 'Google');
+      done();
+    });
   });
 
   it('should accept an object with property and query fields', (done) => {
-    const BASE_URL = 'http://metadata.google.internal/computeMetadata/v1';
-    const TYPE = 'type';
-    const PROPERTY = 'property';
     const QUERY = {key: 'value'};
-
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
-
-    retryRequestOverride = (reqOpts, opts, callback) => {
-      assert.deepEqual(reqOpts, {
-        uri: BASE_URL + '/' + TYPE + '/' + PROPERTY,
-        headers: {'Metadata-Flavor': 'Google'},
-        qs: QUERY
-      });
-      callback(null, VALID_RESPONSE);
-    };
-
-    getMetadata({property: PROPERTY, qs: QUERY}, done);
+    nock(HOST).get(`${PATH}/${TYPE}/${PROPERTY}`).query(QUERY).reply(200, {});
+    getMetadata({property: PROPERTY, params: QUERY}, (err, res) => {
+      assert.equal(res!.config.params, QUERY);
+      assert.equal(res!.config.headers['Metadata-Flavor'], 'Google');
+      assert.equal(res!.config.url, `${BASE_URL}/${TYPE}/${PROPERTY}`);
+      done();
+    });
   });
 
   it('should extend the request options', (done) => {
-    const BASE_URL = 'http://metadata.google.internal/computeMetadata/v1';
-    const TYPE = 'type';
-    const PROPERTY = 'property';
-
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
-
-    retryRequestOverride = (reqOpts, opts, callback) => {
-      assert.deepEqual(reqOpts, {
-        uri: BASE_URL + '/' + TYPE + '/' + PROPERTY,
-        headers: {'Metadata-Flavor': 'Google', 'Custom-Header': 'Custom'}
-      });
-      callback(null, VALID_RESPONSE);
-    };
-
     const options = {property: PROPERTY, headers: {'Custom-Header': 'Custom'}};
-
-    const originalOptions = extend(true, {}, options);
-
-    getMetadata(options, (err) => {
+    const originalOptions = Object.assign({}, options);
+    nock(HOST).get(`${PATH}/${TYPE}/${PROPERTY}`).reply(200, {});
+    getMetadata(options, (err, res) => {
+      assert.equal(res!.config.url, `${BASE_URL}/${TYPE}/${PROPERTY}`);
+      // assert.equal(res!.config.headers['Metadata-Flavor'], 'Google');
+      assert.equal(res!.config.headers['Custom-Header'], 'Custom');
       assert.ifError(err);
       assert.deepEqual(options, originalOptions);  // wasn't modified
       done();
@@ -122,32 +70,17 @@ describe('gcpMetadata', () => {
   });
 
   it('should return the request error', (done) => {
-    const ERROR = Object.assign(new Error('fake error'), {code: 'ETEST'});
-    const TYPE = 'type';
-
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
-
-    retryRequestOverride = (reqOpts, opts, callback) => {
-      callback(ERROR);
-    };
-
-    getMetadata((err) => {
-      // TSC fails if code isn't defined on err.
-      assert.ok(err && typeof (err.code) === 'string');
-
-      assert.strictEqual(err, ERROR);
+    nock(HOST).get(`${PATH}/${TYPE}`).reply(500, {});
+    getMetadata(err => {
+      assert.strictEqual(err!.message, 'Unsuccessful response status code');
       done();
     });
   });
 
   it('should return error when res is empty', (done) => {
-    const TYPE = 'type';
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
-
-    retryRequestOverride = (reqOpts, opts, callback) => {
-      callback(null, null);
-    };
-
+    nock(HOST).get(`${PATH}/${TYPE}`).reply(200, null);
     getMetadata((err) => {
       assert(err instanceof Error);
       done();
@@ -155,13 +88,8 @@ describe('gcpMetadata', () => {
   });
 
   it('should return error when flavor header is incorrect', (done) => {
-    const TYPE = 'type';
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
-
-    retryRequestOverride = (reqOpts, opts, callback) => {
-      callback(null, {headers: {'Metadata-Flavor': 'Hazelnut'}});
-    };
-
+    nock(HOST).get(PATH).reply(200, {}, {'Metadata-Flavor': 'Hazelnut'});
     getMetadata((err) => {
       assert(err instanceof Error);
       done();
@@ -169,13 +97,8 @@ describe('gcpMetadata', () => {
   });
 
   it('should return error if statusCode is not 200', (done) => {
-    const TYPE = 'type';
     const getMetadata = gcpMetadata._buildMetadataAccessor(TYPE);
-
-    retryRequestOverride = (reqOpts, opts, callback) => {
-      callback(null, {headers: {'Metadata-Flavor': 'Google'}, statusCode: 418});
-    };
-
+    nock(HOST).get(PATH).reply(418, {}, {'Metadata-Flavor': 'Google'});
     getMetadata((err) => {
       assert(err instanceof Error);
       done();

